File: 871-minimum-number-of-refueling-stops/871-minimum-number-of-refueling-stops.java
class Solution {
    public int minRefuelStops(int target, int startFuel, int[][] stations) {
        if(startFuel >= target) return 0;
        PriorityQueue<Integer> pq=new PriorityQueue<>(Collections.reverseOrder());
        int i=0,n=stations.length,cnt=0;
        while(true) {
            if(startFuel >= target) return cnt;
            int max=0;
            while(i<n&&stations[i][0]<=startFuel) {
                pq.add(stations[i][1]);
                i++;
            }
            cnt++;
            if(pq.isEmpty()) break;
            startFuel+=pq.poll();
            
        }
        return -1;
    }
}

File: 871-minimum-number-of-refueling-stops/871-minimum-number-of-refueling-stops.js
var minRefuelStops = function(target, startFuel, stations) {
    let dp = new Array(stations.length + 1).fill(0);
    dp[0] = startFuel;
    for (let i = 0; i < stations.length; i++) {
        for (let t = i; t >=0 && dp[t] >= stations[i][0]; t--) {
            dp[t + 1] = Math.max(dp[t + 1], dp[t] + stations[i][1])
        }
    }
    for (let i = 0; i <= stations.length; i++) {
        if (dp[i] >= target) return i;
    }
    
    return -1;
};